### PR TITLE
change to SciMLOperators

### PR DESCRIPTION
**Pull request by vpuri3**
https://github.com/SciML/SciMLOperators.jl/issues/55

### CODE REVIEW

**Comment by vpuri3** [comment]
vector diagonal preconditioner should be updated as follows:
```julia
            s = rand(n)
            #Pl, Pr = DiagonalOperator(s), LinearSolve.InvPreconditioner(Diagonal(s))
            Pl = DiagonalOperator(s)
            Pr = inv(Pl)

            Pl = cache_operator(Pl, x)
            Pr = cache_operator(Pr, x)```

**Comment by vpuri3** [comment]
this is good

**Comment by ChrisRackauckas** [comment]
Tests fail (needs formatting, and downstream tests)

File: lib/LinearSolveCUDA/test/runtests.jl [hunk 1]
```diff
@@ -42,6 +42,9 @@ function test_interface(alg, prob1, prob2)
     return
 end
 
+# TODO - test with SciMLOperators.MatrixOperator
+# TODO - test LinearSolveCUDA on GPU machine
+
```
Comment (ChrisRackauckas) [inline lib/LinearSolveCUDA/test/runtests.jl:47]: This runs on CUDA with CI

File: src/LinearSolve.jl [hunk 2]
```diff
@@ -24,6 +24,12 @@ import IterativeSolvers
 using Reexport
 @reexport using SciMLBase
 
+function isidentity(A)
+    return (A === LinearAlgebra.I) |
+           (A isa IterativeSolvers.Identity) |
+           (A isa SciMLOperators.IdentityOperator)
```
Comment (ChrisRackauckas) [inline src/LinearSolve.jl:30]:
```suggestion
    return (A === LinearAlgebra.I) ||
           (A isa IterativeSolvers.Identity) ||
           (A isa SciMLOperators.IdentityOperator)
```

File: src/LinearSolve.jl [hunk 3]
```diff
@@ -36,24 +42,26 @@ needs_concrete_A(alg::AbstractKrylovSubspaceMethod) = false
...
+
 const IS_OPENBLAS = Ref(true)
 isopenblas() = IS_OPENBLAS[]
 
 export LUFactorization, SVDFactorization, QRFactorization, GenericFactorization,
        GenericLUFactorization, SimpleLUFactorization, RFLUFactorization,
        UMFPACKFactorization, KLUFactorization, FastLUFactorization, FastQRFactorization
 
-export LinearSolveFunction
+export LinearSolveFunction, ApplyLdiv, ApplyLdiv!
```
Comment (ChrisRackauckas) [inline src/LinearSolve.jl:64]: ?

File: src/solve_function.jl [hunk 4]
```diff
@@ -13,3 +12,23 @@ function SciMLBase.solve(cache::LinearCache, alg::LinearSolveFunction,
...
+
+struct ApplyLdiv <: AbstractSolveFunction end
+function SciMLBase.solve(cache::LinearCache, alg::ApplyLdiv, args...; kwargs...)
+    @unpack A, b, u = cache
+
+    v = A \ b
+    copy!(u, v)
+
+    return SciMLBase.build_linear_solution(alg, cache.u, nothing, cache)
+end
```
Comment (ChrisRackauckas) [inline src/solve_function.jl:24]: oh, the naming is weird. `DirectLdiv(inplace = true)`?

File: test/basictests.jl [hunk 5]
```diff
@@ -191,53 +192,6 @@ end
...
-            P = LinearSolve.ComposePreconditioner(P1, P2)
-
-            # ComposePreconditioner
-            ldiv!(y, P, x)
-            @test y ≈ ldiv!(P2, ldiv!(P1, x))
-            y .= x
-            ldiv!(P, x)
-            @test x ≈ ldiv!(P2, ldiv!(P1, y))
-        end
-    end
```
Comment (ChrisRackauckas) [inline test/basictests.jl:239]: why were these deleted?

File: docs/src/solvers/solvers.md [hunk 6]
```diff
@@ -27,6 +27,12 @@ factorization methods if a lower tolerance of the solution is required.
 Krylov.jl generally outperforms IterativeSolvers.jl and KrylovKit.jl, and is compatible
 with CPUs and GPUs, and thus is the generally preferred form for Krylov methods.
 
+Many linear system have custom inverses defined. If `Base.\`, or `LinearAlgebra.ldiv!` are
+overloaded for `A`, then the user can choose the algorithms `ApplyLdiv`, or `ApplyLdiv!` to
+use that. Alternatively, `LinearSolve` will automatically use `Base.\` or
+`LinearAlgebra.ldiv!` if `SciMLOperators.has_ldiv(A)` or `SciMLOperators.has_ldiv!(A)`
+evaluates to `true`.
```
Comment (ChrisRackauckas) [inline docs/src/solvers/solvers.md:34]: I wouldn't recommend this. Normally some pre-factor will still be faster. Most types should probably define this. What's a counter example?

File: src/LinearSolve.jl [hunk 7]
```diff
@@ -36,24 +42,26 @@ needs_concrete_A(alg::AbstractKrylovSubspaceMethod) = false
...
 include("simplelu.jl")
 include("iterative_wrappers.jl")
-include("preconditioners.jl")
 include("solve_function.jl")
 include("default.jl")
 include("init.jl")
 
+# deprecate preconditioner interface in favor of SciMLOperators
+@deprecate ComposePreconditioner SciMLOperators.ComposedOperator
+@deprecate InvPreconditioner SciMLOperators.InvertedOperator
```
Comment (ChrisRackauckas) [inline src/LinearSolve.jl:55]: ahh

File: lib/LinearSolveCUDA/test/runtests.jl (hunk 1 above)
Comment (vpuri3) [reply lib/LinearSolveCUDA/test/runtests.jl:47]: got it

File: src/solve_function.jl (hunk 4 above)
Comment (vpuri3) [reply src/solve_function.jl:24]: that works haha

**Comment by vpuri3** [comment]
do deprecations as follows:

```julia
const DiffEqArrayOperator = MatrixOperator

function DiffEqArrayOperator(args...)
@warn "SciMLBase.DiffEqArrayOperator is deprecated.
Use SciMLOperators.MatrixOperator instead"

MatrixOperator(args...)
end
```

**Comment by vpuri3** [comment]
might have to keep `ComposePreconditioner` since OrdinaryDiffEq uses that

**Comment by ChrisRackauckas** [comment]
It's fine to remove if the deprecation is correct.

**Comment by vpuri3** [comment]
make https://github.com/JuliaSmoothOptimizers/Krylov.jl/issues/612#issuecomment-1227391841 part of this

File: Project.toml [hunk 8]
```diff
@@ -16,7 +16,9 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 RecursiveFactorization = "f2c3362d-daeb-58d1-803e-2bc74f2840b4"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
 SciMLBase = "0bca4576-84f4-4d90-8ffe-ffa030f20462"
+SciMLOperators = "c0aeaf25-5076-4817-a8d5-81caf7dfa961"
 Setfield = "efcf1570-3423-57d1-acb7-fd33fddbac46"
+SnoopPrecompile = "66db9d55-30c0-4569-8b51-7e840670fc0c"
```
Comment (ChrisRackauckas) [inline Project.toml:21]: Why is this in this PR?

File: docs/src/advanced/custom.md [hunk 9]
```diff
@@ -36,7 +37,7 @@ The inputs to the function are as follows:
 The function `my_linsolve` must accept the above specified arguments, and return
 the solution, `u`. As memory for `u` is already allocated, the user may choose
 to modify `u` in place as follows:
-```julia
+```@example advanced1
```
Comment (ChrisRackauckas) [inline docs/src/advanced/custom.md:40]: seems unrelated?

File: docs/src/basics/FAQ.md [hunk 10]
```diff
@@ -2,36 +2,78 @@
 
 Ask more questions.
 
+## How is LinearSolve.jl compared to just using normal \, i.e. A\b?
+
+Check out [this video from JuliaCon 2022](https://www.youtube.com/watch?v=JWI34_w-yYw) which goes
+into detail on how and why LinearSolve.jl is able to be a more general and efficient interface.
+
```
Comment (ChrisRackauckas) [inline docs/src/basics/FAQ.md:9]: Rebase issues?

**Comment by vpuri3** [comment]
yes lotsa rebase issues ://. I have half a mind of abandoning this branch and starting anew
